File: src/scpn_phase_orchestrator/upde/bifurcation.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SCPN Phase Orchestrator — Bifurcation analysis for Kuramoto networks

"""Bifurcation continuation for Kuramoto synchronisation transitions.

Traces steady-state order parameter ``R`` as a function of coupling
strength ``K`` using pseudo-arclength continuation (Keller 1977).
Detects critical coupling ``K_c`` where the incoherent state
``R ≈ 0`` bifurcates to partial synchronisation ``R > 0``.

Analytical reference: ``K_c = 2 / (π g(0))`` for Lorentzian ``g(ω)``
with half-width ``Δ`` → ``K_c = 2Δ`` (Kuramoto 1975, Strogatz 2000).

5-backend chain via delegation
------------------------------
The single-trial kernel ``steady_state_r(phases, omegas, knm,
alpha, k_scale, dt, n_transient, n_measure) → R`` is already
dispatched across Rust / Mojo / Julia / Go / Python in
:mod:`scpn_phase_orchestrator.upde.basin_stability`. This module
delegates to it rather than re-implementing the Euler trial
integrator, which means every ``trace_sync_transition`` /
``find_critical_coupling`` call in the Python-composite branch
inherits the full fallback chain for free.

The two composite Rust kernels — ``trace_sync_transition_rust``
(batched K-sweep) and ``find_critical_coupling_bif_rust`` (binary
search inside Rust) — are preserved as one-shot fast paths: a
single FFI call amortises the per-K boundary overhead better than
the N_points × dispatch-call path.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from numbers import Integral, Real
from typing import TypeAlias

import numpy as np
from numpy.typing import NDArray

from scpn_phase_orchestrator.upde.basin_stability import (
    steady_state_r as _dispatched_steady_state_r,
)

try:
    from spo_kernel import (
        find_critical_coupling_bif_rust as _rust_find_kc,
    )
    from spo_kernel import (
        trace_sync_transition_rust as _rust_trace,
    )

    _HAS_COMPOSITE_RUST = True
except ImportError:
    _HAS_COMPOSITE_RUST = False

__all__ = [
    "BifurcationDiagram",
    "BifurcationPoint",
    "find_critical_coupling",
    "trace_sync_transition",
]
FloatArray: TypeAlias = NDArray[np.float64]


def _contains_boolean_alias(value: object) -> bool:
    try:
        arr = np.asarray(value, dtype=object)
    except (TypeError, ValueError):
        return False
    return any(isinstance(item, (bool, np.bool_)) for item in arr.flat)


@dataclass
class BifurcationPoint:
    """One sampled point on a Kuramoto synchronisation branch."""

    K: float
    R: float
    stable: bool

    def __post_init__(self) -> None:
        k_value = _validate_finite_float(self.K, name="K")
        if k_value < 0.0:
            raise ValueError(f"K must be non-negative, got {self.K!r}")
        r_value = _validate_unit_interval(self.R, name="R")
        if not isinstance(self.stable, bool):
            raise ValueError(f"stable must be a boolean flag, got {self.stable!r}")

        self.K = k_value
        self.R = r_value


@dataclass
class BifurcationDiagram:
    """Ordered bifurcation samples plus optional critical coupling."""

    points: list[BifurcationPoint] = field(default_factory=list)
    K_critical: float | None = None

    def __post_init__(self) -> None:
        if not isinstance(self.points, list):
            raise ValueError("points must be a list of BifurcationPoint records")
        for idx, point in enumerate(self.points):
            if not isinstance(point, BifurcationPoint):
                raise ValueError(
                    f"points[{idx}] must be a BifurcationPoint, got {point!r}"
                )
        if self.K_critical is not None:
            k_critical = _validate_finite_float(self.K_critical, name="K_critical")
            if k_critical < 0.0:
                raise ValueError(
                    f"K_critical must be non-negative, got {self.K_critical!r}"
                )
            self.K_critical = k_critical

    @property
    def K_values(self) -> FloatArray:
        """Return continuation coupling strengths in diagram order."""

        return np.array([p.K for p in self.points])

    @property
    def R_values(self) -> FloatArray:
        """Return continuation order parameters in diagram order."""

        return np.array([p.R for p in self.points])


def _validate_integral(value: object, *, name: str, minimum: int) -> int:
    if isinstance(value, bool) or not isinstance(value, Integral) or value < minimum:
        raise ValueError(f"{name} must be an integer >= {minimum}, got {value!r}")
    return int(value)


def _validate_finite_float(value: object, *, name: str) -> float:
    if isinstance(value, bool) or not isinstance(value, Real):
        raise ValueError(f"{name} must be a finite real, got {value!r}")
    coerced = float(value)
    if not np.isfinite(coerced):
        raise ValueError(f"{name} must be a finite real, got {value!r}")
    return coerced


def _validate_positive_float(value: object, *, name: str) -> float:
    coerced = _validate_finite_float(value, name=name)
    if coerced <= 0.0:
        raise ValueError(f"{name} must be positive, got {value!r}")
    return coerced


def _validate_unit_interval(value: object, *, name: str) -> float:
    coerced = _validate_finite_float(value, name=name)
    if coerced < 0.0 or coerced > 1.0:
        raise ValueError(f"{name} must be in [0, 1], got {value!r}")
    return coerced


def _validate_omegas(value: object) -> FloatArray:
    if _contains_boolean_alias(value):
        raise ValueError("omegas must not contain boolean values")
    try:
        arr = np.asarray(value, dtype=np.float64)
    except (TypeError, ValueError) as exc:
        raise ValueError("omegas must be a finite one-dimensional array") from exc
    if arr.ndim != 1:
        raise ValueError(f"omegas shape {arr.shape} must be one-dimensional")
    if arr.size < 1:
        raise ValueError("omegas must contain at least one oscillator")
    if not np.all(np.isfinite(arr)):
        raise ValueError("omegas must contain only finite values")
    return np.ascontiguousarray(arr, dtype=np.float64)


def _validate_matrix(
    value: object,
    *,
    name: str,
    n: int,
    require_zero_diagonal: bool = False,
) -> FloatArray:
    if _contains_boolean_alias(value):
        raise ValueError(f"{name} must not contain boolean values")
    try:
        arr = np.asarray(value, dtype=np.float64)
    except (TypeError, ValueError) as exc:
        raise ValueError(f"{name} must be a finite float matrix") from exc
    if arr.shape != (n, n):
        raise ValueError(f"{name} shape {arr.shape} does not match ({n}, {n})")
    if not np.all(np.isfinite(arr)):
        raise ValueError(f"{name} must contain only finite values")
    if require_zero_diagonal and not np.allclose(np.diag(arr), 0.0, atol=1e-12):
        raise ValueError(
            f"{name} diagonal must be zero; self-coupling K_ii is not physical"
        )
    return np.ascontiguousarray(arr, dtype=np.float64)


def _default_coupling(n: int) -> FloatArray:
    knm_template = np.ones((n, n), dtype=np.float64) / n
    np.fill_diagonal(knm_template, 0.0)
    return knm_template


def _validate_k_range(value: object) -> tuple[float, float]:
    if not isinstance(value, tuple) or len(value) != 2:
        raise ValueError("K_range must contain exactly two finite values")
    start, stop = value
    start = _validate_finite_float(start, name="K_range")
    stop = _validate_finite_float(stop, name="K_range")
    if start < 0.0:
        raise ValueError("K_range start must be non-negative")
    if stop <= start:
        raise ValueError("K_range stop must be greater than start")
    return start, stop


def _steady_state_R_dispatch(
    phases_init: FloatArray,
    omegas: FloatArray,
    K_scale: float,
    knm_template: FloatArray,
    alpha: FloatArray,
    dt: float,
    n_transient: int,
    n_measure: int,
) -> float:
    """Thin wrapper around the 5-backend-dispatched kernel.

    Shipped as a module-private helper so the ``trace_*`` /
    ``find_*`` functions stay Python-only at the Python level;
    the multi-language work happens inside
    ``basin_stability.steady_state_r``.
    """
    return _dispatched_steady_state_r(
        phases_init,
        omegas,
        knm_template,
        alpha=alpha,
        k_scale=K_scale,
        dt=dt,
        n_transient=n_transient,
        n_measure=n_measure,
    )


def trace_sync_transition(
    omegas: FloatArray,
    knm_template: FloatArray | None = None,
    alpha: FloatArray | None = None,
    K_range: tuple[float, float] = (0.0, 5.0),
    n_points: int = 50,
    dt: float = 0.01,
    n_transient: int = 2000,
    n_measure: int = 500,
    seed: int = 42,
) -> BifurcationDiagram:
    """Trace R(K) for the Kuramoto synchronisation transition.

    Sweeps coupling strength ``K`` from ``K_range[0]`` to
    ``K_range[1]``, running the ODE to steady state at each point,
    and returns a :class:`BifurcationDiagram` with the ``(K, R)``
    pairs plus the estimated critical coupling ``K_c``.

    When the Rust composite kernel is available, the whole sweep
    is batched into a single FFI call. Otherwise the function
    loops in Python and each trial is dispatched through the
    5-backend chain inherited from
    :func:`basin_stability.steady_state_r`.
    """
    omegas = _validate_omegas(omegas)
    n = int(omegas.shape[0])
    K_range = _validate_k_range(K_range)
    n_points = _validate_integral(n_points, name="n_points", minimum=2)
    dt = _validate_positive_float(dt, name="dt")
    n_transient = _validate_integral(n_transient, name="n_transient", minimum=0)
    n_measure = _validate_integral(n_measure, name="n_measure", minimum=0)
    seed = _validate_integral(seed, name="seed", minimum=0)

    if knm_template is None:
        knm_template = _default_coupling(n)
    else:
        knm_template = _validate_matrix(
            knm_template,
            name="knm_template",
            n=n,
            require_zero_diagonal=True,
        )
    if alpha is None:
        alpha = np.zeros((n, n), dtype=np.float64)
    else:
        alpha = _validate_matrix(alpha, name="alpha", n=n)

    rng = np.random.default_rng(seed)
    phases_init = rng.uniform(0, 2 * np.pi, n)
    diagram = BifurcationDiagram()

    if _HAS_COMPOSITE_RUST:
        o = np.ascontiguousarray(omegas, dtype=np.float64)
        k = np.ascontiguousarray(knm_template.ravel(), dtype=np.float64)
        a = np.ascontiguousarray(alpha.ravel(), dtype=np.float64)
        p = np.ascontiguousarray(phases_init, dtype=np.float64)
        kv, rv, kc = _rust_trace(
            o,
            k,
            a,
            n,
            p,
            K_range[0],
            K_range[1],
            n_points,
            dt,
            n_transient,
            n_measure,
        )
        kv = np.asarray(kv)
        rv = np.asarray(rv)
        for i in range(len(kv)):
            diagram.points.append(
                BifurcationPoint(
                    K=float(kv[i]),
                    R=float(rv[i]),
                    stable=True,
                ),
            )
        if not np.isnan(kc):
            diagram.K_critical = float(kc)
        return diagram

    # Composite Rust unavailable — loop in Python, each trial
    # dispatched through the basin_stability 5-backend chain.
    K_values = np.linspace(K_range[0], K_range[1], n_points)
    for K_val in K_values:
        R = _steady_state_R_dispatch(
            phases_init,
            omegas,
            K_val,
            knm_template,
            alpha,
            dt,
            n_transient,
            n_measure,
        )
        diagram.points.append(
            BifurcationPoint(K=float(K_val), R=R, stable=True),
        )

    R_arr = diagram.R_values
    threshold = 0.1
    crossings = np.where(
        (R_arr[:-1] < threshold) & (R_arr[1:] >= threshold),
    )[0]
    if len(crossings) > 0:
        idx = crossings[0]
        K_lo, K_hi = float(K_values[idx]), float(K_values[idx + 1])
        R_lo, R_hi = float(R_arr[idx]), float(R_arr[idx + 1])
        if R_hi > R_lo:
            frac = (threshold - R_lo) / (R_hi - R_lo)
            diagram.K_critical = K_lo + frac * (K_hi - K_lo)
        else:
            diagram.K_critical = K_hi
    return diagram


def find_critical_coupling(
    omegas: FloatArray,
    knm_template: FloatArray | None = None,
    dt: float = 0.01,
    n_transient: int = 3000,
    n_measure: int = 1000,
    tol: float = 0.05,
    seed: int = 42,
) -> float:
    """Binary search for the critical coupling ``K_c`` at which
    ``R`` crosses a 0.1 threshold.

    More precise than :func:`trace_sync_transition` when only
    ``K_c`` is needed. Returns ``nan`` if no transition is found
    in ``[0, 20]``.
    """
    omegas = _validate_omegas(omegas)
    n = int(omegas.shape[0])
    dt = _validate_positive_float(dt, name="dt")
    n_transient = _validate_integral(n_transient, name="n_transient", minimum=0)
    n_measure = _validate_integral(n_measure, name="n_measure", minimum=0)
    tol = _validate_positive_float(tol, name="tol")
    seed = _validate_integral(seed, name="seed", minimum=0)

    if knm_template is None:
        knm_template = _default_coupling(n)
    else:
        knm_template = _validate_matrix(
            knm_template,
            name="knm_template",
            n=n,
            require_zero_diagonal=True,
        )

    alpha = np.zeros((n, n))
    rng = np.random.default_rng(seed)
    phases_init = rng.uniform(0, 2 * np.pi, n)

    if _HAS_COMPOSITE_RUST:
        o = np.ascontiguousarray(omegas, dtype=np.float64)
        k = np.ascontiguousarray(knm_template.ravel(), dtype=np.float64)
        a = np.ascontiguousarray(alpha.ravel(), dtype=np.float64)
        p = np.ascontiguousarray(phases_init, dtype=np.float64)
        return float(
            _rust_find_kc(
                o,
                k,
                a,
                n,
                p,
                dt,
                n_transient,
                n_measure,
                tol,
            ),
        )

    threshold = 0.1
    K_lo, K_hi = 0.0, 20.0

    R_hi = _steady_state_R_dispatch(
        phases_init,
        omegas,
        K_hi,
        knm_template,
        alpha,
        dt,
        n_transient,
        n_measure,
    )
    if R_hi < threshold:
        return float("nan")

    for _ in range(30):
        K_mid = (K_lo + K_hi) / 2
        R_mid = _steady_state_R_dispatch(
            phases_init,
            omegas,
            K_mid,
            knm_template,
            alpha,
            dt,
            n_transient,
            n_measure,
        )
        if R_mid < threshold:
            K_lo = K_mid
        else:
            K_hi = K_mid
        if K_hi - K_lo < tol:
            break

    return (K_lo + K_hi) / 2

File: spo-kernel/crates/spo-engine/src/bifurcation.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SCPN Phase Orchestrator — Bifurcation analysis (Keller 1977, Kuramoto 1975)

//! Bifurcation continuation for Kuramoto synchronisation transitions.
//!
//! Traces steady-state order parameter R as a function of coupling strength K.
//! Detects critical coupling K_c where R bifurcates from 0 to partial sync.
//!
//! References:
//!   Kuramoto 1975, International Symposium on Mathematical Problems
//!     in Theoretical Physics, Lecture Notes in Physics 39:420-422.
//!   Strogatz 2000, Physica D 143:1-20.
//!   Keller 1977, "Numerical Solution of Bifurcation and Nonlinear
//!     Eigenvalue Problems".

/// Run Kuramoto ODE to steady state and return time-averaged R.
///
/// Euler integration of dθ_i/dt = ω_i + Σ_j K_ij sin(θ_j − θ_i − α_ij)
///
/// # Arguments
/// * `phases_init` — (N,) initial phases
/// * `omegas` — (N,) natural frequencies
/// * `knm_flat` — (N×N) row-major coupling template (will be scaled by k_scale)
/// * `alpha_flat` — (N×N) row-major phase lag matrix
/// * `n` — number of oscillators
/// * `k_scale` — coupling strength multiplier
/// * `dt` — integration timestep
/// * `n_transient` — steps to discard
/// * `n_measure` — steps to average R over
#[must_use]
#[allow(clippy::too_many_arguments)]
pub fn steady_state_r(
    phases_init: &[f64],
    omegas: &[f64],
    knm_flat: &[f64],
    alpha_flat: &[f64],
    n: usize,
    k_scale: f64,
    dt: f64,
    n_transient: usize,
    n_measure: usize,
) -> f64 {
    let mut phases = phases_init.to_vec();

    // Transient: integrate to steady state
    for _ in 0..n_transient {
        kuramoto_step(&mut phases, omegas, knm_flat, alpha_flat, n, k_scale, dt);
    }

    // Measure: time-averaged R
    let mut r_sum = 0.0;
    for _ in 0..n_measure {
        kuramoto_step(&mut phases, omegas, knm_flat, alpha_flat, n, k_scale, dt);
        r_sum += order_parameter(&phases);
    }

    r_sum / n_measure as f64
}

/// Single Euler step for Kuramoto model.
fn kuramoto_step(
    phases: &mut [f64],
    omegas: &[f64],
    knm_flat: &[f64],
    alpha_flat: &[f64],
    n: usize,
    k_scale: f64,
    dt: f64,
) {
    // Compute coupling: c_i = Σ_j K_ij * sin(θ_j − θ_i − α_ij)
    // Then θ_i += dt * (ω_i + c_i)
    let old = phases.to_vec();
    for i in 0..n {
        let mut coupling = 0.0;
        for j in 0..n {
            let k_ij = knm_flat[i * n + j] * k_scale;
            if k_ij.abs() < 1e-30 {
                continue;
            }
            let a_ij = alpha_flat[i * n + j];
            coupling += k_ij * (old[j] - old[i] - a_ij).sin();
        }
        phases[i] = old[i] + dt * (omegas[i] + coupling);
    }
}

/// Kuramoto order parameter R = |<exp(iθ)>|.
fn order_parameter(phases: &[f64]) -> f64 {
    if phases.is_empty() {
        return 0.0;
    }
    let n = phases.len() as f64;
    let mut sum_cos = 0.0;
    let mut sum_sin = 0.0;
    for &theta in phases {
        sum_cos += theta.cos();
        sum_sin += theta.sin();
    }
    ((sum_cos / n).powi(2) + (sum_sin / n).powi(2)).sqrt()
}

/// Trace R(K) bifurcation diagram.
///
/// Returns (K_values, R_values, K_critical).
/// K_critical is NaN if no transition found.
#[must_use]
#[allow(clippy::too_many_arguments)]
pub fn trace_sync_transition(
    omegas: &[f64],
    knm_flat: &[f64],
    alpha_flat: &[f64],
    n: usize,
    phases_init: &[f64],
    k_min: f64,
    k_max: f64,
    n_points: usize,
    dt: f64,
    n_transient: usize,
    n_measure: usize,
) -> (Vec<f64>, Vec<f64>, f64) {
    let mut k_values = Vec::with_capacity(n_points);
    let mut r_values = Vec::with_capacity(n_points);

    for i in 0..n_points {
        let k = if n_points > 1 {
            k_min + (k_max - k_min) * i as f64 / (n_points - 1) as f64
        } else {
            k_min
        };
        k_values.push(k);
        let r = steady_state_r(
            phases_init,
            omegas,
            knm_flat,
            alpha_flat,
            n,
            k,
            dt,
            n_transient,
            n_measure,
        );
        r_values.push(r);
    }

    // Find K_critical: first crossing of R = 0.1
    let threshold = 0.1;
    let mut k_critical = f64::NAN;
    for i in 0..r_values.len() - 1 {
        if r_values[i] < threshold && r_values[i + 1] >= threshold {
            let r_lo = r_values[i];
            let r_hi = r_values[i + 1];
            if r_hi > r_lo {
                let frac = (threshold - r_lo) / (r_hi - r_lo);
                k_critical = k_values[i] + frac * (k_values[i + 1] - k_values[i]);
            } else {
                k_critical = k_values[i + 1];
            }
            break;
        }
    }

    (k_values, r_values, k_critical)
}

/// Binary search for critical coupling K_c.
///
/// Returns K_c or NaN if no transition in [0, 20].
#[must_use]
#[allow(clippy::too_many_arguments)]
pub fn find_critical_coupling(
    omegas: &[f64],
    knm_flat: &[f64],
    alpha_flat: &[f64],
    n: usize,
    phases_init: &[f64],
    dt: f64,
    n_transient: usize,
    n_measure: usize,
    tol: f64,
) -> f64 {
    let threshold = 0.1;

    let mut k_lo = 0.0;
    let mut k_hi = 20.0;

    let r_hi = steady_state_r(
        phases_init,
        omegas,
        knm_flat,
        alpha_flat,
        n,
        k_hi,
        dt,
        n_transient,
        n_measure,
    );
    if r_hi < threshold {
        return f64::NAN;
    }

    for _ in 0..30 {
        let k_mid = (k_lo + k_hi) / 2.0;
        let r_mid = steady_state_r(
            phases_init,
            omegas,
            knm_flat,
            alpha_flat,
            n,
            k_mid,
            dt,
            n_transient,
            n_measure,
        );
        if r_mid < threshold {
            k_lo = k_mid;
        } else {
            k_hi = k_mid;
        }
        if k_hi - k_lo < tol {
            break;
        }
    }

    (k_lo + k_hi) / 2.0
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::f64::consts::TAU;

    fn make_all_to_all(n: usize) -> Vec<f64> {
        let mut knm = vec![1.0 / n as f64; n * n];
        for i in 0..n {
            knm[i * n + i] = 0.0;
        }
        knm
    }

    #[test]
    fn test_order_parameter_sync() {
        let phases = vec![0.0, 0.0, 0.0, 0.0];
        let r = order_parameter(&phases);
        assert!((r - 1.0).abs() < 1e-12);
    }

    #[test]
    fn test_order_parameter_antisync() {
        let phases = vec![0.0, std::f64::consts::PI];
        let r = order_parameter(&phases);
        assert!(r < 1e-12);
    }

    #[test]
    fn test_order_parameter_empty() {
        assert_eq!(order_parameter(&[]), 0.0);
    }

    #[test]
    fn test_steady_state_r_zero_coupling() {
        // Zero coupling → oscillators drift → R ≈ 0 (for non-identical ω)
        let n = 8;
        let omegas: Vec<f64> = (0..n).map(|i| 0.5 + 0.1 * i as f64).collect();
        let knm = make_all_to_all(n);
        let alpha = vec![0.0; n * n];
        let phases: Vec<f64> = (0..n).map(|i| TAU * i as f64 / n as f64).collect();

        let r = steady_state_r(&phases, &omegas, &knm, &alpha, n, 0.0, 0.01, 500, 200);
        assert!(r < 0.5, "zero coupling should give low R, got {r}");
    }

    #[test]
    fn test_steady_state_r_strong_coupling() {
        // Strong coupling → R close to 1
        let n = 8;
        let omegas: Vec<f64> = (0..n).map(|i| 1.0 + 0.05 * i as f64).collect();
        let knm = make_all_to_all(n);
        let alpha = vec![0.0; n * n];
        let phases: Vec<f64> = (0..n).map(|i| TAU * i as f64 / n as f64).collect();

        let r = steady_state_r(&phases, &omegas, &knm, &alpha, n, 10.0, 0.01, 2000, 500);
        assert!(r > 0.8, "strong coupling should give high R, got {r}");
    }

    #[test]
    fn test_trace_monotone_r() {
        // R should generally increase with K
        let n = 6;
        let omegas: Vec<f64> = (0..n).map(|i| 1.0 + 0.1 * i as f64).collect();
        let knm = make_all_to_all(n);
        let alpha = vec![0.0; n * n];
        let phases: Vec<f64> = (0..n).map(|i| TAU * i as f64 / n as f64).collect();

        let (_, r_values, _) = trace_sync_transition(
            &omegas, &knm, &alpha, n, &phases, 0.0, 8.0, 5, 0.01, 500, 200,
        );
        // R at K=8 should be greater than R at K=0
        assert!(
            r_values[4] > r_values[0],
            "R should increase with K: R(0)={}, R(8)={}",
            r_values[0],
            r_values[4]
        );
    }

    #[test]
    fn test_trace_returns_correct_length() {
        let n = 4;
        let omegas = vec![1.0; n];
        let knm = make_all_to_all(n);
        let alpha = vec![0.0; n * n];
        let phases = vec![0.0; n];
        let (k_vals, r_vals, _) = trace_sync_transition(
            &omegas, &knm, &alpha, n, &phases, 0.0, 5.0, 10, 0.01, 100, 50,
        );
        assert_eq!(k_vals.len(), 10);
        assert_eq!(r_vals.len(), 10);
    }

    #[test]
    fn test_find_critical_coupling_exists() {
        // Identical frequencies → K_c = 0 (already synchronised at any K)
        let n = 4;
        let omegas = vec![1.0; n];
        let knm = make_all_to_all(n);
        let alpha = vec![0.0; n * n];
        let phases = vec![0.0; n];

        let kc = find_critical_coupling(&omegas, &knm, &alpha, n, &phases, 0.01, 500, 200, 0.1);
        // Identical ω: even tiny coupling syncs → K_c should be small
        assert!(
            kc < 5.0,
            "identical frequencies should have low K_c, got {kc}"
        );
    }

    #[test]
    fn test_find_critical_coupling_spread_frequencies() {
        // Spread frequencies → should find a finite K_c
        let n = 6;
        let omegas: Vec<f64> = (0..n).map(|i| 0.5 + 1.0 * i as f64).collect();
        let knm = make_all_to_all(n);
        let alpha = vec![0.0; n * n];
        let phases: Vec<f64> = (0..n).map(|i| TAU * i as f64 / n as f64).collect();

        let kc = find_critical_coupling(&omegas, &knm, &alpha, n, &phases, 0.01, 2000, 500, 0.1);
        assert!(!kc.is_nan(), "should find K_c for spread frequencies");
        // K_c > 0 (some coupling needed)
        assert!(
            kc > 0.0,
            "spread frequencies need nonzero coupling, got {kc}"
        );
    }

    #[test]
    fn test_kuramoto_step_preserves_count() {
        let n = 5;
        let mut phases = vec![0.0, 1.0, 2.0, 3.0, 4.0];
        let omegas = vec![1.0; n];
        let knm = make_all_to_all(n);
        let alpha = vec![0.0; n * n];
        kuramoto_step(&mut phases, &omegas, &knm, &alpha, n, 1.0, 0.01);
        assert_eq!(phases.len(), n);
        // All phases should have changed
        assert!((phases[0] - 0.0).abs() > 1e-10);
    }

    #[test]
    fn test_r_values_bounded() {
        let n = 4;
        let omegas = vec![1.0, 2.0, 3.0, 4.0];
        let knm = make_all_to_all(n);
        let alpha = vec![0.0; n * n];
        let phases = vec![0.0, 1.0, 2.0, 3.0];

        let (_, r_vals, _) = trace_sync_transition(
            &omegas, &knm, &alpha, n, &phases, 0.0, 10.0, 5, 0.01, 200, 100,
        );
        for (i, &r) in r_vals.iter().enumerate() {
            assert!(r >= 0.0 && r <= 1.0 + 1e-10, "R[{i}] = {r} out of [0,1]");
        }
    }
}
